modules/javafx.graphics/src/main/java/javafx/scene/Node.java

```diff
@@ -8112,55 +8119,85 @@ public final ObjectProperty<InputMethodRequests> inputMethodRequestsProperty() {
+            return Node.this;
         }
 
-        public void notifyListeners() {
-            if (needsChangeEvent) {
-                fireValueChangedEvent();
-                needsChangeEvent = false;
-            }
+        @Override
+        public boolean get() {
+            return value;
         }
 
-        private void markInvalid() {
-            if (valid) {
-                valid = false;
+        public void set(boolean value) {
+            this.value = value;
+        }
 
-                pseudoClassStateChanged(FOCUSED_PSEUDOCLASS_STATE, get());
-                PlatformLogger logger = Logging.getFocusLogger();
-                if (logger.isLoggable(Level.FINE)) {
-                    logger.fine(this + " focused=" + get());
-                }
+        protected boolean notifyListeners() {
+            if (lastNotifiedValue == value) {
+                return false;
+            }
 
-                needsChangeEvent = true;
+            lastNotifiedValue = value;
+            pseudoClassStateChanged(getPseudoClass(), value);
 
-                notifyAccessibleAttributeChanged(AccessibleAttribute.FOCUSED);
+            PlatformLogger logger = Logging.getFocusLogger();
+            if (logger.isLoggable(Level.FINE)) {
+                logger.fine(this + " " + getName() + "=" + get());
             }
-        }
 
-        @Override
-        public boolean get() {
-            valid = true;
-            return value;
+            fireValueChangedEvent();
+            return true;
         }
+    }
 
-        @Override
-        public Object getBean() {
-            return Node.this;
-        }
+    /**
+     * Sets the value of the {@link #focused} and {@link #focusVisible} properties
+     * without firing change events. The value of {@link #focusWithin} is set
+     * by the implementation of {@link #focused}.
+     */
+    final void setFocusQuietly(boolean focused, boolean focusVisible) {
+        this.focused.set(focused);
+        this.focusVisible.set(focused && focusVisible);
+    }
 
-        @Override
-        public String getName() {
-            return "focused";
+    /**
+     * Fires change notifications if the value of {@link #focused}, {@link #focusVisible}
+     * or {@link #focusWithin} has changed. Change notifications for {@link #focusWithin}
+     * are fired on the current node and on all of its parents, if necessary.
+     */
+    final void notifyFocusListeners() {
+        focused.notifyListeners();
+        focusVisible.notifyListeners();
+
+        Node node = this;
+        do {
+            node.focusWithin.notifyListeners();
+            node = node.getParent();
+        } while (node != null);
+    }
+
+    /**
+     * Called when the current node was removed from the scene graph in order to clear
+     * the focus bits of the former parents.
+     */
+    private void updateRemovedParentFocus(Node oldParent) {
+        if (oldParent != null && focusWithin.get()) {
```

Comment:
I see that you aren't calling `this.focusWithin.set(false)`. Is that because in the case of the node that was removed from its parent, it will have its `focused` property set to false (as a result of it being removed from the scene), which will then set its `focusWithin` to false?

Reply:
Yes, that's correct. A node that is removed from a scene graph cannot be focused, which also means that it can't have focusWithin.